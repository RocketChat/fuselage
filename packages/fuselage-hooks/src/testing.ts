import {
  queries,
  Queries,
  RenderHookOptions,
  RenderHookResult,
  renderHook as _renderHook,
} from '@testing-library/react';
import { createElement } from 'react';
import * as ReactDOMClient from 'react-dom';
import { renderToString } from 'react-dom/server';

type RendererableContainer = Element | Document | DocumentFragment;

export function renderHook<
  Result,
  Props,
  Q extends Queries = typeof queries,
  Container extends RendererableContainer = HTMLElement,
  BaseElement extends RendererableContainer = Container,
>(
  render: (initialProps: Props) => Result,
  options?: RenderHookOptions<Props, Q, Container, BaseElement> | undefined,
): RenderHookResult<Result, Props> {
  if (typeof document === 'undefined') {
    let current: Result;
    const TestComponent = () => {
      current = render(options?.initialProps as any);
      return null;
    };

    renderToString(createElement(TestComponent));

    return {
      result: { current: current! },
      rerender: () => undefined,
      unmount: () => undefined,
    };
  }

  if ('createRoot' in ReactDOMClient) return _renderHook(render, options);

  return _renderHook(render, { ...options, legacyRoot: true });
}

export { act } from '@testing-library/react';
