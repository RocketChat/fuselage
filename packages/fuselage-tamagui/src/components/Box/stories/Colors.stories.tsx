import type { Meta, StoryObj } from '@storybook/react'
import { Box } from '../Box'

const meta = {
  title: 'Layout/Box/Colors',
  component: Box,
  tags: ['autodocs'],
  parameters: {
    date: '2025-07-03 12:40:14',
    author: 'Muskan0400'
  }
} satisfies Meta<typeof Box>

export default meta
type Story = StoryObj<typeof Box>

export const SurfaceColors: Story = {
  args: {
    backgroundColor: '$surface',
    children: 'Surface Colors',
    padding: '$4'
  }
}

export const StatusColors: Story = {
  args: {
    backgroundColor: '$status',
    children: 'Status Colors',
    padding: '$4'
  }
}

export const StrokeColors: Story = {
  args: {
    borderColor: '$stroke',
    children: 'Stroke Colors',
    padding: '$4'
  }
}

export const FontColors: Story = {
  args: {
    color: '$font',
    children: 'Font Colors',
    padding: '$4'
  }
}