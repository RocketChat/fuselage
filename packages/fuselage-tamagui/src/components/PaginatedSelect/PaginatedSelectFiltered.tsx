import { styled, Input, Theme } from 'tamagui';
import { useState, useCallback } from 'react';
import type { ComponentProps } from 'react';

import { PaginatedSelect } from './PaginatedSelect';

const FilterInput = styled(Input, {
  flex: 1,
  padding: '$2',
  borderRadius: '$2',
  borderWidth: 1,
  borderColor: '$borderColor',
});

type PaginatedSelectFilteredProps = Omit<ComponentProps<typeof PaginatedSelect>, 'filter' | 'setFilter'> & {
  filter?: string;
  setFilter: (filter: string) => void;
};

export function PaginatedSelectFiltered({
  filter = '',
  setFilter,
  placeholder,
  ...props
}: PaginatedSelectFilteredProps) {
  const [isOpen, setIsOpen] = useState(false);

  const handleFilterChange = useCallback((e: any) => {
    setFilter(e.target.value);
  }, [setFilter]);

  return (
    <Theme name="light">
      <PaginatedSelect
        {...props}
        filter={filter}
        setFilter={setFilter}
        placeholder={undefined}
        renderInput={() => (
          <FilterInput
            value={filter}
            onChange={handleFilterChange}
            placeholder={placeholder}
            onFocus={() => setIsOpen(true)}
            onBlur={() => setTimeout(() => setIsOpen(false), 200)}
          />
        )}
      />
    </Theme>
  );
}

export default PaginatedSelectFiltered;