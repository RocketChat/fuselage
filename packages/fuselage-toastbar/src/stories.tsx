import { Button, Box } from '@rocket.chat/fuselage';
import type { Meta, Story } from '@storybook/react';
import { useState } from 'react';

import { useToastBarDispatch } from './ToastBarContext';

export default {
  title: 'view/ToastBar',
  parameters: {
    layout: 'fullscreen',
    actions: { argTypesRegex: '^on.*' },
  },
} as Meta;

export const ToastBarWithData: Story = () => {
  const [counter, setCounter] = useState(0);
  const dispatchToastMessage = useToastBarDispatch();

  const messageArray = [
    'Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsam nihi Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsam nihi',
    'Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsam nihi Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsam nihi',
    'Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsam nihi Lorem ipsum dolor, sit amet consectetur adipisicing elit. Ipsam nihi',
    'Lorem Ipsum',
    'Lorem ipsum dolor, sit amet consectetur adipisicing elit.',
  ];

  const handleToast = () => {
    dispatchToastMessage({
      type: 'success',
      message: messageArray[counter],
    });

    dispatchToastMessage({
      type: 'error',
      message: messageArray[counter],
      time: 10,
      position: 'bottom-start',
    });

    if (counter === messageArray.length - 1) {
      return setCounter(0);
    }

    return setCounter((prevState) => prevState + 1);
  };

  return (
    <Box
      display='flex'
      alignItems='center'
      justifyContent='center'
      height='100vh'
    >
      <Button primary onClick={handleToast}>
        Dispatch ToastBar
      </Button>
    </Box>
  );
};
