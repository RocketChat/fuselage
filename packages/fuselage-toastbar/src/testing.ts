import type {
  queries,
  Queries,
  RenderOptions,
  RenderResult,
} from '@testing-library/react';
import { render as renderOriginal } from '@testing-library/react';
import type React from 'react';
import type ReactDOMClient from 'react-dom';

type RendererableContainer = ReactDOMClient.Container;
type HydrateableContainer = Parameters<(typeof ReactDOMClient)['hydrate']>[0];

/**
 * Light wrapper around `react-testing-library` to provide a custom render function for transitioning to React 18
 */
export function render<
  Q extends Queries = typeof queries,
  Container extends RendererableContainer | HydrateableContainer = HTMLElement,
  BaseElement extends RendererableContainer | HydrateableContainer = Container,
>(
  ui: React.ReactNode,
  options: RenderOptions<Q, Container, BaseElement>,
): RenderResult<Q, Container, BaseElement>;
export function render(
  ui: React.ReactNode,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult;
export function render(ui: React.ReactNode, options?: any): any {
  return renderOriginal(ui, {
    legacyRoot: true,
    ...options,
  });
}
