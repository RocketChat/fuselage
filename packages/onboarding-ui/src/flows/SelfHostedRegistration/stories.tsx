import { Box, Callout } from '@rocket.chat/fuselage';
import type { Meta, Story } from '@storybook/react';
import { useState } from 'react';

import type { AdminInfoPayload } from '../../forms/AdminInfoForm/AdminInfoForm';
import type { OrganizationInfoPayload } from '../../forms/OrganizationInfoForm/OrganizationInfoForm';
import type { RegisteredServerPayload } from '../../forms/RegisteredServerForm/RegisteredServerForm';
import AdminInfoPage from '../../pages/AdminInfoPage';
import AwaitingConfirmationPage from '../../pages/AwaitingConfirmationPage';
import ConfirmationProcessPage from '../../pages/ConfirmationProcessPage';
import EmailConfirmedPage from '../../pages/EmailConfirmedPage';
import OrganizationInfoPage from '../../pages/OrganizationInfoPage';
import RegisteredServerPage from '../../pages/RegisteredServerPage';
import StandaloneServerPage from '../../pages/StandaloneServerPage';
import {
  countryOptions,
  logSubmit,
  organizationIndustryOptions,
  organizationSizeOptions,
  organizationTypes,
  validateEmail,
  validatePassword,
  validateUsername,
} from './mocks';

export default {
  title: 'flows/Self-Hosted Registration',
  parameters: {
    layout: 'fullscreen',
    actions: { argTypesRegex: '^on.*' },
  },
} as Meta;

export const SelfHostedRegistration: Story = () => {
  const [path, navigateTo] =
    useState<`/${
      | 'admin-info'
      | 'org-info'
      | 'register-server'
      | 'standalone-confirmation'
      | 'cloud-email'
      | 'awaiting'
      | 'home'
      | 'email'
      | 'confirmation-progress'
      | 'email-confirmed'}`>('/admin-info');

  const [adminInfo, setAdminInfo] =
    useState<Omit<AdminInfoPayload, 'password'>>();

  const [organizationInfo, setOrganizationInfo] =
    useState<OrganizationInfoPayload>();

  const [serverRegistration, setServerRegistration] = useState<{
    updates?: boolean;
    agreement?: boolean;
    cloudAccountEmail?: string;
    securityCode?: string;
  }>();

  const handleAdminInfoSubmit = logSubmit((data: AdminInfoPayload) => {
    setAdminInfo(data);
    navigateTo('/org-info');
  });

  const handleOrganizationInfoSubmit = logSubmit(
    (data: OrganizationInfoPayload) => {
      setOrganizationInfo(data);
      navigateTo('/register-server');
    }
  );

  const handleRegisterServerSubmit = logSubmit(
    (data: RegisteredServerPayload) => {
      setServerRegistration((serverRegistration) => ({
        ...serverRegistration,
        updates: data.updates,
        agreement: data.agreement,
        cloudAccountEmail: data.email,
        securityCode: 'Funny Tortoise In The Hat',
      }));
      navigateTo('/awaiting');
    }
  );

  if (path === '/admin-info') {
    return (
      <AdminInfoPage
        currentStep={1}
        stepCount={4}
        passwordRulesHint=''
        validateUsername={validateUsername}
        validateEmail={validateEmail}
        validatePassword={validatePassword}
        initialValues={adminInfo}
        onSubmit={handleAdminInfoSubmit}
      />
    );
  }

  if (path === '/org-info') {
    return (
      <OrganizationInfoPage
        currentStep={2}
        stepCount={4}
        organizationTypeOptions={organizationTypes}
        organizationIndustryOptions={organizationIndustryOptions}
        organizationSizeOptions={organizationSizeOptions}
        countryOptions={countryOptions}
        initialValues={organizationInfo}
        onBackButtonClick={() => navigateTo('/admin-info')}
        onSubmit={handleOrganizationInfoSubmit}
      />
    );
  }

  if (path === '/register-server') {
    return (
      <RegisteredServerPage
        currentStep={3}
        stepCount={4}
        initialValues={{
          ...(serverRegistration?.updates && {
            updates: serverRegistration?.updates,
          }),
          ...(serverRegistration?.agreement && {
            agreement: serverRegistration?.agreement,
          }),
        }}
        onBackButtonClick={() => navigateTo('/org-info')}
        onSubmit={handleRegisterServerSubmit}
        onClickContinue={() => navigateTo('/standalone-confirmation')}
      />
    );
  }

  if (path === '/standalone-confirmation') {
    return (
      <StandaloneServerPage
        currentStep={4}
        stepCount={4}
        initialValues={{}}
        onBackButtonClick={() => navigateTo('/register-server')}
        onSubmit={() => navigateTo('/home')}
      />
    );
  }

  if (path === '/awaiting') {
    if (!serverRegistration?.cloudAccountEmail) {
      throw new Error('missing cloud account email');
    }

    if (!serverRegistration?.securityCode) {
      throw new Error('missing verification code');
    }

    setTimeout(() => {
      navigateTo('/confirmation-progress');
    }, 5000);

    return (
      <AwaitingConfirmationPage
        emailAddress={serverRegistration.cloudAccountEmail}
        securityCode={serverRegistration.securityCode}
        onChangeEmailRequest={() => navigateTo('/admin-info')}
        onResendEmailRequest={() => undefined}
      />
    );
  }

  if (path === '/confirmation-progress') {
    setTimeout(() => {
      navigateTo('/email-confirmed');
    }, 3000);

    return <ConfirmationProcessPage />;
  }

  if (path === '/email-confirmed') {
    return <EmailConfirmedPage />;
  }

  if (path === '/home') {
    return (
      <Box
        width='100vw'
        height='100vh'
        display='flex'
        justifyContent='center'
        alignItems='center'
      >
        <Callout type='success'>This is the home of the workspace.</Callout>
      </Box>
    );
  }

  throw new Error('invalid path');
};
SelfHostedRegistration.storyName = 'Self-Hosted Registration';
