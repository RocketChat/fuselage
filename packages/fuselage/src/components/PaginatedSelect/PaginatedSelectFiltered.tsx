import { useMutableCallback } from '@rocket.chat/fuselage-hooks';
import type { FormEvent, Ref } from 'react';
import React, { useMemo, forwardRef } from 'react';

import { InputBox } from '../InputBox';
import type { PaginatedSelectProps } from './PaginatedSelect';
import PaginatedSelect from './PaginatedSelect';

type PaginatedSelectFilteredProps = Omit<PaginatedSelectProps, 'setFilter'> & {
  setFilter: (value: string | undefined | number) => void;
};

const PaginatedSelectFiltered = ({
  filter,
  setFilter,
  options,
  placeholder,
  ...props
}: PaginatedSelectFilteredProps) => {
  const anchor = useMemo(
    () =>
      // eslint-disable-next-line react/no-multi-comp
      forwardRef(function AnchorComponent(
        { filter, onChange: _onChange, ...props }: PaginatedSelectFilteredProps,
        ref: Ref<HTMLInputElement>
      ) {
        return (
          <InputBox.Input
            mi='x4'
            flexGrow={1}
            className='rcx-select__focus'
            ref={ref}
            placeholder={placeholder}
            value={filter}
            onChange={useMutableCallback((e: FormEvent<HTMLInputElement>) => {
              setFilter(e.currentTarget.value);
            })}
            {...props}
            rcx-input-box--undecorated
          />
        );
      }),
    [placeholder, setFilter]
  );

  return (
    <PaginatedSelect
      placeholder={undefined}
      filter={filter}
      options={options}
      {...props}
      anchor={anchor}
    />
  );
};

export default PaginatedSelectFiltered;
