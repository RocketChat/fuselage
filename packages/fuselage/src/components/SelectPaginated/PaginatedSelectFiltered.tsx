import { useMutableCallback } from '@rocket.chat/fuselage-hooks';
import type { FormEvent, Ref } from 'react';
import React, { useMemo, forwardRef } from 'react';

import { InputBox } from '../InputBox';
import type { PaginatedSelectProps } from './PaginatedSelect';
import { PaginatedSelect } from './PaginatedSelect';

type PaginatedSelectFilteredProps = Omit<PaginatedSelectProps, 'setFilter'> & {
  setFilter: (value: string | undefined | number) => void;
};

export const PaginatedSelectFiltered = ({
  filter,
  setFilter,
  options,
  placeholder,
  ...props
}: PaginatedSelectFilteredProps) => {
  const anchor = useMemo(
    () =>
      forwardRef(
        (
          {
            filter,
            onChange: _onChange,
            ...props
          }: PaginatedSelectFilteredProps,
          ref: Ref<HTMLInputElement>
        ) => (
          <InputBox.Input
            mi='x4'
            flexGrow={1}
            className='rcx-select__focus'
            ref={ref}
            placeholder={placeholder}
            value={filter}
            onChange={useMutableCallback((e: FormEvent<HTMLInputElement>) => {
              setFilter(e.currentTarget.value);
            })}
            {...props}
            rcx-input-box--undecorated
          />
        )
      ),
    [placeholder, setFilter]
  );

  return (
    <PaginatedSelect
      placeholder={undefined}
      filter={filter}
      options={options}
      {...props}
      anchor={anchor}
    />
  );
};
