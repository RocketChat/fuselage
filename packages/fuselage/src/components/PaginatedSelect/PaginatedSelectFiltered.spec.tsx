import { composeStories } from '@storybook/react-webpack5';
import { screen } from '@testing-library/dom';
import { axe } from 'jest-axe';

import { render } from '../../testing';

import { PaginatedSelectFiltered } from './PaginatedSelectFiltered';
import * as stories from './PaginatedSelectFiltered.stories';

const testCases = Object.values(composeStories(stories)).map((Story) => [
  Story.storyName || 'Story',
  Story,
]);

describe('[PaginatedSelectFiltered Component]', () => {
  test.each(testCases)(
    `renders %s without crashing`,
    async (_storyname, Story) => {
      const tree = render(<Story />);
      expect(tree.baseElement).toMatchSnapshot();
    },
  );

  test.each(testCases)(
    '%s should have no a11y violations',
    async (_storyname, Story) => {
      const { container } = render(<Story />);

      const results = await axe(container);
      expect(results).toHaveNoViolations();
    },
  );

  test('should render placeholder when provided', async () => {
    const hiddenClass = 'rcx-select__wrapper--hidden';
    const placeholder = 'Select an option...';
    const defaultProps = {
      setFilter: jest.fn(),
      onChange: jest.fn(),
      options: [{ value: 'item1', label: `Item #1` }],
    };

    const { rerender } = render(
      <PaginatedSelectFiltered {...defaultProps} placeholder={placeholder} />,
    );

    expect(screen.getByRole('textbox')).toHaveAttribute(
      'placeholder',
      placeholder,
    );

    // Necessary due to styles not being properly computed in the test environment
    expect(screen.getByRole('textbox').parentElement).not.toHaveClass(
      hiddenClass,
    );

    rerender(<PaginatedSelectFiltered {...defaultProps} value='item1' />);

    expect(screen.getByRole('textbox')).not.toHaveAttribute(
      'placeholder',
      placeholder,
    );

    // Necessary due to styles not being properly computed in the test environment
    expect(screen.getByRole('textbox').parentElement).toHaveClass(hiddenClass);
  });
});
