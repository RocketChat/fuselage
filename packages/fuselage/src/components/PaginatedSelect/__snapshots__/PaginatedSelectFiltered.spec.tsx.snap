// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`[PaginatedSelectFiltered Component] renders Disabled without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-select disabled rcx-css-5gkzmp"
      disabled=""
    >
      <div
        class="rcx-box rcx-box--full rcx-select__wrapper rcx-css-bpb89k"
      >
        <input
          aria-haspopup="listbox"
          class="rcx-box rcx-box--full rcx-box--animated rcx-input-box--undecorated rcx-input-box rcx-select__focus rcx-css-1l0s473"
          disabled=""
          placeholder="Placeholder here..."
          value=""
        />
        <div
          class="rcx-box rcx-box--full rcx-select__addon rcx-css-x7bl3q rcx-css-6bg1ps"
        >
          <i
            aria-hidden="true"
            class="rcx-box rcx-box--full rcx-icon--name-chevron-down rcx-icon rcx-css-4pvxx3"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`[PaginatedSelectFiltered Component] renders Errored without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-select invalid rcx-css-5gkzmp"
    >
      <div
        class="rcx-box rcx-box--full rcx-select__wrapper rcx-css-bpb89k"
      >
        <input
          aria-haspopup="listbox"
          class="rcx-box rcx-box--full rcx-box--animated rcx-input-box--undecorated rcx-input-box rcx-select__focus rcx-css-1l0s473"
          placeholder="Placeholder here..."
          value=""
        />
        <div
          class="rcx-box rcx-box--full rcx-select__addon rcx-css-x7bl3q rcx-css-6bg1ps"
        >
          <i
            aria-hidden="true"
            class="rcx-box rcx-box--full rcx-icon--name-chevron-down rcx-icon rcx-css-4pvxx3"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`[PaginatedSelectFiltered Component] renders Normal without crashing 1`] = `
<body>
  <div>
    <div
      class="rcx-box rcx-box--full rcx-select rcx-css-5gkzmp"
    >
      <div
        class="rcx-box rcx-box--full rcx-select__wrapper rcx-css-bpb89k"
      >
        <input
          aria-haspopup="listbox"
          class="rcx-box rcx-box--full rcx-box--animated rcx-input-box--undecorated rcx-input-box rcx-select__focus rcx-css-1l0s473"
          placeholder="Placeholder here..."
          value=""
        />
        <div
          class="rcx-box rcx-box--full rcx-select__addon rcx-css-x7bl3q rcx-css-6bg1ps"
        >
          <i
            aria-hidden="true"
            class="rcx-box rcx-box--full rcx-icon--name-chevron-down rcx-icon rcx-css-4pvxx3"
          >
            
          </i>
        </div>
      </div>
    </div>
  </div>
</body>
`;
